src/metabase/driver/drill.clj

```diff
@@ -0,0 +1,245 @@
+                               3)
+                         (hsql/raw "INTERVAL '3' MONTH")))
+    ;;:quarter (hsql/call :date_trunc_quarter (hx/->timestamp expr))
+    :quarter-of-year (hx/->integer
+                      (hsql/call :ceil
+                                 (hx// (hsql/call :extract
+                                                  "month"
+                                                  (hx/->timestamp expr))
+                                       3.0)))
+    :year (hsql/call :extract "year" (hx/->timestamp expr))))
+
+(defn- humanize-connection-error-message [message]
+  (condp re-matches message
+    #"^FATAL: database \".*\" does not exist$"
+    (driver/connection-error-messages :database-name-incorrect)
+
+    #"^No suitable driver found for.*$"
+    (driver/connection-error-messages :invalid-hostname)
+
+    #"^Connection refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections.$"
+    (driver/connection-error-messages :cannot-connect-check-host-and-port)
+
+    #"^FATAL: .*$" ; all other FATAL messages: strip off the 'FATAL' part, capitalize, and add a period
+    (let [[_ message] (re-matches #"^FATAL: (.*$)" message)]
+      (str (s/capitalize message) \.))
+
+    #".*" ; default
+    message))
+
+(defn- describe-database [driver database]
+  {:tables (with-open [conn (jdbc/get-connection (sql/db->jdbc-connection-spec database))]
+             (set (for [result (jdbc/query {:connection conn} ["select table_schema, table_name from INFORMATION_SCHEMA.`VIEWS` union select table_schema, table_name from INFORMATION_SCHEMA.`TABLES` where table_type='TABLE'"])]
+                    {:name (:table_name result)
+                     :schema (:table_schema result)})))})
+
+(defn- describe-table [driver database table]
+  (with-open [conn (jdbc/get-connection (sql/db->jdbc-connection-spec database))]
+    {:name (:name table)
+     :schema (:schema table)
+     :fields (set (for [result (jdbc/query {:connection conn}
+                                           [(str "select column_name, data_type from INFORMATION_SCHEMA.COLUMNS where table_name='" (:name table) "'")])]
+                    {:name (:column_name result)
+                     :base-type (column->base-type (keyword (:data_type result)))}))}))
+
+(defprotocol ^:private IUnprepare
+  (unprepare-arg ^String [this]))
+
+(extend-protocol IUnprepare
+  nil     (unprepare-arg [this] "NULL")
+  String  (unprepare-arg [this] (str \' (s/replace this "'" "''") \')) ; escape single-quotes
+  Boolean (unprepare-arg [this] (if this "TRUE" "FALSE"))
+  Number  (unprepare-arg [this] (str this))
+  Date    (unprepare-arg [this] (first (hsql/format
+                                        (hsql/call :to_timestamp
+                                                   (hx/literal (u/date->iso-8601 this))
+                                                   (hx/literal "YYYY-MM-dd''T''HH:mm:ss.SSSZ"))))))
+
+(defn- unprepare
```

Comment:
Is this needed for Spark SQL/Drill? It's there for BigQuery because BigQuery isn't JDBC-based and doesn't have prepared statements. But if SparkSQL/Drill has a JDBC driver shouldn't prepared statements work normally?